### ISSUE TITLE
[BUG] Rancher can't start on air-gapped environments

### ISSUE THREAD

**Issue by JacieChao** [rancher/rancher#43779]
**Rancher Server Setup**
- Rancher version: v2.8.0
- Installation option (Docker install/Helm Chart):
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc):

Docker install, Helm Chart install on RKE1/RKE2/K3s on air-gapped environments are not work

**Information about the Cluster**
- Kubernetes version: v1.27.7+rke2r2
- Cluster Type (Local/Downstream): Local

**Describe the bug**

Rancher keep restart no matter using Docker install or deploy by helm on RKE1/RKE2/K3s on the air-gapped environment. 
The logs show errors with fetch images for ui-extensions from GitHub URL.
```
2023/12/13 04:27:20 [INFO] Waiting for initial data to be populated
2023/12/13 04:27:20 [ERROR] failed to call leader func: failed to add management data: failed to fetch images from extensions: Get "https://api.github.com/repos/rancher/ui-plugin-charts/releases": dial tcp 20.27.177.116:443: i/o timeout
2023/12/13 04:27:22 [INFO] Waiting for initial data to be populated
2023/12/13 04:27:24 [INFO] Waiting for initial data to be populated
2023/12/13 04:27:25 [INFO] Reconciling GlobalRoles
2023/12/13 04:27:25 [INFO] Reconciling RoleTemplates
2023/12/13 04:27:26 [INFO] Waiting for initial data to be populated
```

**To Reproduce**

- Prepare rancher images by v2.8.0 rancher-image.txt and save all images to harbor.
- Prepare private registry configuration for RKE2
- Install v1.27.7+rke2r2 on the air-gapped environment by the [instruction](https://docs.rke2.io/install/airgap#rke2-installsh-script-install) and waiting for the server running
- Deploy rancher with the following command.
```
helm install rancher \
    ./rancher-2.8.0.tgz \
    --namespace cattle-system \
    --set hostname=x.x.x.x \
    --set rancherImage=x.x.x.x/rancher/rancher \
    --set ingress.tls.source=secret \
    --set privateCA=true \
    --set replicas=1 \
    --set systemDefaultRegistry=x.x.x.x \
    --set useBundledSystemChart=true
```
- The rancher pod keeps restarting.
```
rancher-8f5f4f677-d58mm            0/1     Running     1 (61s ago)   4m1s
```

**Result**

**Expected Result**

The rancher pod can start in the air-gapped environment.

**Screenshots**

**Additional context**

Deep dive into this issue, I found this issue is relative to this [PR](https://github.com/rancher/rancher/pull/42565). Rancher stuck in [add catalog data](https://github.com/rancher/rancher/blob/release/v2.8/pkg/data/management/catalog_data.go#L211) which uses the same [GetImage](https://github.com/rancher/rancher/blob/release/v2.8/pkg/image/extensions.go#L25) function modified by the PR. In [FetchExtensionImages](https://github.com/rancher/rancher/blob/release/v2.8/pkg/image/resolve.go#L89) uses the [hard-coded Github URL](https://github.com/rancher/rancher/blob/release/v2.8/pkg/image/extensions.go#L25) to fetch the image which will return error in the air-gapped environment. Rancher will restart with this error and the error logs shows the failed messages.

**Comment by glongzh** [comment]
encountered the same problem

**Comment by bmdepesa** [comment]
@glongzh @JacieChao were these fresh installations or upgraded environments?

**Comment by JacieChao** [comment]
@bmdepesa 
This issue only occurred on a fresh installation. 

Validated upgrade from v2.7.9 to v2.8.0 rancher server can start properly. But the log keeps ouputing
```
[ERROR] failed to call leader func: failed to add management data: failed to fetch images from extensions: Get "https://api.github.com/repos/rancher/ui-plugin-charts/releases": dial tcp 20.27.177.116:443: i/o timeout
```

**Comment by snasovich** [comment]
@JacieChao , thank you for reporting this issue and doing deeper dive into the issue cause. Since it seems to be caused by UI specific changes, moving to that team.
@gaktive @nwmac , FYI. Since it's backend specific I think we should keep this issue in `rancher/rancher` instead of moving to `rancher/dashboard` but feel free to change as needed.

**Comment by igomez06** [comment]
Verified this on a fresh install. With three server nodes (cp, master, etcd), 1 agent node all air gapped, and it came up fine. This is what my helm command looked like
`helm upgrade --install --create-namespace --set hostname=<hostname> --set rancherImage=<registry hostname>/rancher/rancher --set systemDefaultRegistry=<registry hostname> --version 2.8.0 --devel --wait -n cattle-system rancher rancher-latest/rancher `

**Comment by snasovich** [comment]
@igomez06 , based on code quoted by @JacieChao it looks like having `SystemCatalog` set to `bundled` (which is set via `--set useBundledSystemChart=true`) is important for reproducing the issue. I don't see it in your command and wondering if you could retest with that option set?

**Comment by igomez06** [comment]
Sorry about that @snasovich and @JacieChao I reran my command and set `-set useBundledSystemChart=true` and I was able to reproduce the issue. My rancher pods keep restarting. Here is my event messages

```
Events:
  Type     Reason     Age                    From               Message
  ----     ------     ----                   ----               -------
  Normal   Scheduled  20m                    default-scheduler  Successfully assigned cattle-system/rancher-7dcd9ddb75-26x9v to ip-192-168-18-230
  Normal   Pulling    20m                    kubelet            Pulling image "<registry-url>/rancher/rancher:v2.8.0"
  Normal   Pulled     19m                    kubelet            Successfully pulled image "<registry-url>/rancher/rancher:v2.8.0" in 43.503336876s (43.503350396s including waiting)
  Normal   Created    17m (x2 over 19m)      kubelet            Created container rancher
  Normal   Started    17m (x2 over 19m)      kubelet            Started container rancher
  Normal   Killing    17m                    kubelet            Container rancher failed liveness probe, will be restarted
  Normal   Pulled     17m                    kubelet            Container image "<registry-url>/rancher/rancher:v2.8.0" already present on machine
  Warning  Unhealthy  16m (x4 over 18m)      kubelet            Liveness probe failed: Get "[http://10.42.54.131:80/healthz](http://10.42.54.131/healthz)": dial tcp 10.42.54.131:80: connect: connection refused
  Warning  Unhealthy  10m (x37 over 19m)     kubelet            Readiness probe failed: Get "[http://10.42.54.131:80/healthz](http://10.42.54.131/healthz)": dial tcp 10.42.54.131:80: connect: connection refused
  Warning  BackOff    5m18s (x2 over 5m19s)  kubelet            Back-off restarting failed container rancher in pod rancher-7dcd9ddb75-26x9v_cattle-system(431d7f11-7ca7-4449-a2c3-0377c38c8fbc)
```

**Comment by snasovich** [comment]
@rohitsakala , this issue is to track addressing it specifically in `release/v2.8` branch while https://github.com/rancher/rancher/issues/43910 created from it is for `release/v2.8.1` branch from which we're releasing `v2.8.1` from.
So, I'm changing the milestone on this one back to `v2.8-Next1` (or actually more correct "next regular 2.8.x release" is `v2.8.3` milestone - don't ask why).

Assiging to @prabalsharma and moving to "Waiting for Rancher RC" as this being airgap issue I believe it requires RC so we will need to wait for `v2.8.3-rc1`.

FYI @Jono-SUSE-Rancher

**Comment by Ghostwritten** [comment]
helm offline  deploy rancher v2.7.5: [https://ghostwritten.blog.csdn.net/article/details/132695704](https://ghostwritten.blog.csdn.net/article/details/132695704) no promblem!

but helm offline deploy rancher 2.8.0 :[https://ghostwritten.blog.csdn.net/article/details/135499028](https://ghostwritten.blog.csdn.net/article/details/135499028),have the same issue:

```

[INFO] Waiting for initial data to be populated
2024/01/18 14:25:42 [ERROR] failed to call leader func: failed to add management data: failed to fetch images from extensions: Get "https://api.github.com/repos/rancher/ui-plugin-charts/releases": dial tcp 20.205.243.168:443: connect: connection refused
```

**Comment by Jono-SUSE-Rancher** [comment]
Added to the project board and moved it to blocked until we can get a 2.8x-next1-rc1

**Comment by markusewalker** [comment]
Validated that this is addressed on `v2.8.3-rc4`.

**ENVIRONMENT DETAILS**
- Rancher version: `v2.8.3-rc4`
- Rancher install: Helm

**TEST RESULTS**
|#|Scenario|Test Case|Result
|---|---|---|---|
|1|Setup Rancher air-gap w/bundled charts|Create a Rancher v2.8.3-rc4 setup with Helm option `useBundledCharts`, validate the Rancher pods startup|:white_check_mark:|

**VALIDATION STEPS**
1. Created private registry and loaded up images from `rancher-images.txt` for `v2.8.3-rc4`.
2. Spun up separate 3 nodes and created a HA RKE setup.
3. Ran `helm install` command specifying my environment details and ensured I had `useBundledChart` enabled.
4. Validated rancher pods are in a status of `Running` as expected.
